### PR TITLE
Fix list of users who can be assigned to a ticket from Kanban

### PR DESCRIPTION
**Pull request by Lainow**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !30496

Rewrite of the form used to add one or more actors to a ticket. The form is now generated by ajax and reuses the model present on the main page of an ItilObject.
**Before**
image

**After**
image

### CODE REVIEW

**Comment by cconard96** [comment]
I think if we rely on `used` to store every ID that cannot be selected, there will be issues with performance on databases with a lot of regular users. Maybe it is better to find a way to replace the itemtype and item dropdowns with combined ones like we use in the ticket itself. Those use the `Dropdown::getDropdownActors` function on the server-side which accepts a user permission to check instead.

**Review by cedric-anne — DISMISSED** [review]
Filtering should be based on same rights than the filtering done in the ticket form. It uses the following filtering: https://github.com/glpi-project/glpi/blob/cff3997dbdac96730c97960f1743ddd7722e2327/src/User.php#L4253-L4260

Also, passing ids of all users that cannot be assigned in the `used` option may result in performances issues for GLPI instances that have thousand of users with a simplified interface.

**Review by Lainow — COMMENTED** [review]
I don't think this is the best solution, but I can't seem to use getDropdownActors and another method to get the same result as with the variable "used". I decide to set all self-service users in the "used" var before checking user rights.

**Comment by cedric-anne** [comment]
> I think if we rely on `used` to store every ID that cannot be selected, there will be issues with performance on databases with a lot of regular users. Maybe it is better to find a way to replace the itemtype and item dropdowns with combined ones like we use in the ticket itself. Those use the `Dropdown::getDropdownActors` function on the server-side which accepts a user permission to check instead.

I agree, we should use the same logic as in `Dropdown::getDropdownActors()` for tickets kanban. Maybe we can use this method directly. The actors dropdown in projects Kanban should also be checked, I guess it has the same lack of filtering.

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Have you tried using the `right` parameter ?
You can see the possibles values in `User::getSqlSearchResult()` (and add a new value if there isn't what you need).

**Comment by Lainow** [comment]
To filter the list of available actors that can be associated with a ticket in Kanban, I've started implementing the CommonItilObject actor addition model. But we don't know how to retrieve the ID of the selected ticket.

I've pushed the functional part of my solution, but we need to find a way of retrieving the ID, and we also need to adapt the method of adding a team member to a ticket from within Kanban.

**Comment by stonebuzz** [comment]
@cconard96 
we have difficulty handling the kanban implementation / behavior.

With the latest commit @Lainow tries to reuse ```components/itilobject/actors/main.html.twig``` which is in charge of this display

image

This works, but you'll notice that we're using the hardcoded ```Ticket``` (which could be replaced by  ```supported_itemtype```) with ID ```1```, because we can't find a way to retrieve the related ticket ID (from Kanban card)

NB:  to use ```components/itilobject/actors/main.html.twig``` we need a real ```item``` and related ```itiltemplate``` to work properly

If we succeed, only the following elements will remain to be adapted
- The method for adding actors (as the dropdown list allows multiple selection)
- A small display bug, which breaks the ```select2``` if the modal is closed and reopened.

If you have an idea / solution to find this Identifier that would be great

++

**Comment by cconard96** [comment]
I was thinking that the content of the team modal could be retrieved via AJAX like I did with the item information panel rather than using the elements added inside the kanban.html.twig file.

The entire for loop in the Twig template that adds `template` tags could be removed from kanban.html.twig.
In `ajax/kanban.php` another case would be added to handle "load_team_modal_content" which returns a rendered twig template containing the content which was originally in the `template` element in the main twig template. The AJAX code could be given the item's type and ID to be able to initialize the dropdowns properly.
In the `Kanban.js`, the `showTeamModal` function would be changed to get the content from the AJAX endpoint instead of the `template` element. This function already identifies the item type and ID based on the given card element.

File: ajax/kanban.php [hunk 1]
```diff
@@ -89,7 +89,7 @@
             return;
         }
     }
-    if (in_array($action, ['add_teammember'])) {
+    if (in_array($action, ['load_teammember_form', 'add_teammember'])) {
```
Comment (stonebuzz) [inline ajax/kanban.php:92]: Is it really necessary to check the action ```load_teammember_form``` here?

As verified/used later (L301)

File: templates/components/itilobject/actors/field.html.twig [hunk 2]
```diff
@@ -33,7 +33,8 @@
 
 {% set rand = random() %}
 
-{% set actors = item.getActorsForType(actortypeint, params) %}
+{% set load_selected_actors = params['load_selected_actors'] ?? true %}
```
Comment (stonebuzz) [inline templates/components/itilobject/actors/field.html.twig:36]:
```suggestion
{% set load_actors = params['load_actors'] ?? true %}
```

File: ajax/kanban.php [hunk 3]
```diff
@@ -297,6 +298,33 @@
...
+            'full_width' => true,
+            'fields_template' => $itiltemplate,
+            'add_field_class' => 'col-sm-12',
+        ];
+        TemplateRenderer::getInstance()->display('components/itilobject/actors/main.html.twig', [
+            'item' => $item,
+            'entities_id' => 0,
+            'canupdate' => true,
+            'canassign' => true,
+            'params' => $item->fields + ['load_selected_actors' => false],
```
Comment (stonebuzz) [inline ajax/kanban.php:319]:
```suggestion
            'params' => $item->fields + ['load_actors' => false],
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

File: ajax/kanban.php (hunk 1 above)
Comment (Lainow) [reply ajax/kanban.php:92]: This is used here to load the item

**Review by stonebuzz — APPROVED** [review]

**Review by cconard96 — APPROVED** [review]

**Review by AdrienClairembault — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by cedric-anne — CHANGES_REQUESTED** [review]
It breaks the feature in the Project Kanban context.

**Comment by stonebuzz** [comment]
> It breaks the feature in the Project Kanban context.

```
[2024-01-18 15:25:34] glpiphplog.CRITICAL:   *** Uncaught Exception Error: Call to undefined method Project::getITILTemplateToUse() in /GLPI/10.0-bugfixes/ajax/kanban.php at line 303  
```

```Project``` doesn't inherit from ```CommonITILObject```  which contains the ```getITILTemplateToUse``` method

It doesn't seem possible to use a ```Trait,``` as this function is linked to / uses the value of the ```itilcategories_id``` field.
Field, which does not exist in the ```Project```

Reusing ```components/itilobject/actors/main.html.twig``` from ```Projects``` is not possible either, as ```Projects``` do not use this twig template (the object contains its own method for managing ```showTeam()```) and that the data and design are not the same (actors in the left panel for ```CommonITILObject``` and in a dedicated tab for Projects). 

In addition to ```User```, ```Supplier``` and ```Group```, project teams also contain ```Contacts```.

Two solutions (IMHO)

1. Refactor ```Project``` class to inherit ```CommonITILObject``` (and use ```ITILTemplate``` like ```Ticket```, ```Change``` and ```Problem```) 
2. Use a dedicated method / TWIG template for update actors from the kaban
3. Keep the old method for ```project``` only

**Comment by cedric-anne** [comment]
IMHO, the `Kanban` class should implement a default `getTeamMemberForm()` method that renders a default form (based on the current behaviour) and this method should be overriden in the `CommonITILObject` class to render a form that contains filterd actors.

**Comment by cedric-anne** [comment]
I put this PR back to draft. Please remove the draft status once it will be ready for a review.

File: ajax/kanban.php [hunk 4]
```diff
@@ -278,9 +278,14 @@
     }
 } else if (($_POST['action'] ?? null) === 'add_teammember') {
     $checkParams(['itemtype_teammember', 'items_id_teammember']);
-    $item->addTeamMember($_POST['itemtype_teammember'], (int) $_POST['items_id_teammember'], [
-        'role'   => (int) $_POST['role']
-    ]);
+    if ($_POST['itemtype'] == 'Project') {
```
Comment (stonebuzz) [inline ajax/kanban.php:281]:
```suggestion
    if ($_POST['itemtype'] == Project::class) {
```

**Review by stonebuzz — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Lint is failing for `templates/components/kanban/teammember.html.twig`

**Review by AdrienClairembault — APPROVED** [review]

File: js/modules/Kanban/Kanban.js [hunk 5]
```diff
@@ -1040,7 +1040,17 @@ class GLPIKanbanRights {
...
+            // Issue is noticed when content is injected multiple times. Scripts execute the first time only.
+            const scripts = $(content).find('script');
+            scripts.detach();
             modal.find('.modal-body').html(content);
+            scripts.each(function() {
+                const script = document.createElement('script');
+                script.type = 'text/javascript';
+                script.text = this.innerHTML;
+                modal.find('.modal-body').append(script);
+            });
```
Comment (cedric-anne) [inline js/modules/Kanban/Kanban.js:1053]: This sounds like a hack. What is the initial issue to solve with this precisely?

File: src/Features/Kanban.php [hunk 6]
```diff
@@ -176,4 +180,20 @@ public function getKanbanUrlWithID(int $items_id, bool $full = true): string
...
+
+        TemplateRenderer::getInstance()->display('components/kanban/teammember.html.twig', [
+            'item' => $item,
+            'entities_id' => 0,
+            'canupdate' => true,
+            'canassign' => true,
+            'params' => $item->fields + ['load_actors' => false],
+            'members_types' => $members_types,
+            'canassigntome' => false,
+            'allow_auto_submit' => false,
```
Comment (cedric-anne) [inline src/Features/Kanban.php:196]: Many of these variables are not used in the `components/kanban/teammember.html.twig` template.

File: js/modules/Kanban/Kanban.js (hunk 5 above)
Comment (cconard96) [reply js/modules/Kanban/Kanban.js:1053]: The new team member modal content contains an inline script for the select2 dropdowns. The first time the modal is shown, the inline script is executed as expected. Subsequent injections of the content with the script results in the script not being evaluated again. Maybe there is some other way to fix the issue, but at the time this was the only thing I found.

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
> New methods needs unit tests. I guess changes from "left join" to "inner join" should be tested aswell.

I added some test case in the existing tests for relevant changes.

The new `getTeamMemberForm()` methods are returning an HTML string corresponding to a form. I am not sure it is possible to make a relevant test for this.

**Review by trasher — APPROVED** [review]

**Pull request merged**